File: packages/web/src/algorithms/QC/ruleMissingData.ts
import { clamp } from 'lodash'

import type { AnalysisResultWithClade, NucleotideSubstitution } from 'src/algorithms/types'
import { getQCRuleStatus } from 'src/algorithms/QC/QCRuleStatus'

export interface QCRulesConfigMissingData {
  missingDataThreshold: number
  scoreBias: number
}

export function ruleMissingData(
  { nucleotideComposition }: AnalysisResultWithClade,
  _1: NucleotideSubstitution[],
  { missingDataThreshold, scoreBias }: QCRulesConfigMissingData,
) {
  const totalMissing = nucleotideComposition.N ?? 0

  let score = ((totalMissing - scoreBias) * 100) / missingDataThreshold
  score = clamp(score, 0, Infinity)

  const status = getQCRuleStatus(score)

  return { score, totalMissing, missingDataThreshold: missingDataThreshold + scoreBias, status }
}

export type QCResultMissingData = ReturnType<typeof ruleMissingData>

File: packages/nextclade/src/qc/ruleMissingData.cpp
#include "ruleMissingData.h"

#include <nextclade/nextclade.h>

#include <optional>
#include <vector>

#include "../utils/mapFind.h"
#include "getQcRuleStatus.h"

namespace Nextclade {
  std::optional<QcResultMissingData> ruleMissingData(//
    const AnalysisResult& result,                    //
    const std::vector<NucleotideSubstitution>&,      //
    const QCRulesConfigMissingData& config           //
  ) {
    if (!config.enabled) {
      return {};
    }

    const auto totalMissing = mapFind(result.nucleotideComposition, Nucleotide::N).value_or(0);

    const auto score = std::max(0.0, ((totalMissing - config.scoreBias) * 100) / config.missingDataThreshold);
    const auto& status = getQcRuleStatus(score);

    return QcResultMissingData{
      .score = score,
      .status = status,
      .totalMissing = totalMissing,
      .missingDataThreshold = config.missingDataThreshold + config.scoreBias,// FIXME: Quite confusing. Why is that so?
    };
  }

}// namespace Nextclade
